react/more-react-concepts/reducing-state/reducing-state.md

```diff
@@ -0,0 +1,101 @@
+### Introduction
+
+You have probably heard about reducers before. In this lesson we are going to
+learn _what_ they are, _when_ to use them and _how_ to use them in React with the `useReducer` hook.
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+* What are reducers.
+* When to use reducers.
+* What is the useReducer hook.
+
+### What are reducers?
+
+Reducers are pure functions that take a previous state and an action to return a new state.
+
+The action is an object with a _type_ property describing what the user did. This object can also contain
+any other properties that our reducer might need to produce the new state.
+
+We can define a reducer that updates the state of a counter like this:
+
+~~~js
+function reducer(state, action) {
+  switch (action.type) {
+      case 'incremented_count': {
+        return { count: state.count + 1 }
+        }
+      case 'decremented_count': {
+          return { count: state.count - 1 }
+        }
+      case 'set_count': {
+          return { count: action.value }
+        }
+      default: {
+          throw error('unknown action: ' + action.type);
+        }
+    }
+}
+~~~
+
+Remember, reducers are pure functions so we shouldn't mutate the state.
+
+### When to use reducers?
+
+If a component only needs to update its state in a couple of simple ways, you don't need to use reducers.
+The opposite can be said when a component has become too big, hard to read or debug because of its state logic.
+
+By using reducers, we can separate the state logic and even store it in a different file or directory, leading 
+to smaller components that are easier to read.
+
+Since reducers use _actions_, we can easily track any state-related bugs back to the dispatched action,
+and because our reducers are just pure functions, we can test them in isolation.
+
+### The useReducer hook
+
+React allows us to use reducers in our components through a hook called `useReducer`. This hook takes a reducer function and
+an initial state as arguments, then returns an array with two values: the current state and a `dispatch` function.
+
+This `dispatch` function receives an **action** object as argument, which is passed to our reducer function and 
+the returned value from it is used to update the state.
+
+~~~js
+const [state, dispatch] = useReducer(reducer, { count: 0 });
+
+function handleClick() {
+  dispatch({ type: 'incremented_count' });
+}
+~~~
+
+Similarly to the `set` function in `useState`, React _only updates the state in the next render_ after calling the `dispatch` function.
+Keep in mind that these functions use `Object.is` to determine if the state has changed, if it hasn't, then the component won't re-render.
```

Comment:
> sure :) this is for the dispatch function: https://react.dev/reference/react/useReducer#setstate-caveats
> this is for useState, or more precisely, the 'set' function, I think I should make that change: https://react.dev/reference/react/useState#setstate-caveats
> 
> Should I add one of these resources or both somewhere?

Replying to your comment here.

Thanks for the links. Since the useReducer docs are covered in the assignment section, I don't see the need to link them here as well. Instead we could link to the `Object.is` MDN docs for the learners to refer to. What do you think?
```suggestion
Keep in mind that these functions use [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) to determine if the state has changed, if it hasn't, then the component won't re-render.
```

Reply:
You're absolutely right